### ISSUE TITLE
Odd formatting of type def with many generics

### ISSUE THREAD

**Issue by maltekliemann** [rust-lang/rustfmt#5892]
I'm seeing this behavior in rustfmt 1.6.0-nightly (c469197b 2023-08-22):

```rust
// Before (correct in stable)
type AAAAAAAAAAAAA: BBBBBBBBBBBBBBB<
    CCCCCCCCCCCCCCCCC,
    DDDDDDDDDDDDDDDDD,
    EEEEEEEEEEEEEEEEE,
    FFFFFFFFFFFFFFFFF,
    GGGGGGGGGGGGGGGGG,
    HHHHHHHHHHHHHHHHH,
    IIIIIIIIIIIIIIIII,
>;

// After (hopefully not correct)
type AAAAAAAAAAAAA: BBBBBBBBBBBBBBB<
        CCCCCCCCCCCCCCCCC,
        DDDDDDDDDDDDDDDDD,
        EEEEEEEEEEEEEEEEE,
        FFFFFFFFFFFFFFFFF,
        GGGGGGGGGGGGGGGGG,
        HHHHHHHHHHHHHHHHH,
        IIIIIIIIIIIIIIIII,
    >;
```

It's completely unclear to me what the expected behavior is at this point, but judging by https://github.com/rust-lang/rustfmt/issues/5577 (specifically this: https://github.com/rust-lang/rustfmt/issues/5577#issuecomment-1407489330), this is not what you want.

**Comment by calebcartwright** [comment]
I did have a bit of a panic when I saw this report earlier today, but fortunately, the _default_ behavior is still correct and doesn't have any bearing on stable vs. nightly

Using the nightly-only `version` config option with the non-default value means opting in to formatting improvements, but at the risk of bugs/formatting diffs.

This is absolutely something that needs to be fixed as it's definitely wrong in the v=2 case, but it's not something we need to madly scramble to fix and try to get backported into the 1.72 release :relieved:

**Comment by maltekliemann** [comment]
Oh, sorry, didn't mean to cause a panic by underspecifying the problem. Thanks for taking a look so quickly. :pray:

**Comment by calebcartwright** [comment]
> Oh, sorry, didn't mean to cause a panic by underspecifying the problem. Thanks for taking a look so quickly. 🙏

np, thanks for the report!

**Comment by johnhuichen** [comment]
Currently similar code but for the trait is formatted OK
```rust
trait SomeTrait:
    BBBBBBBBBBBBBBB<
        CCCCCCCCCCCCCCCCC,
        DDDDDDDDDDDDDDDDD,
        EEEEEEEEEEEEEEEEE,
        FFFFFFFFFFFFFFFFF,
        GGGGGGGGGGGGGGGGG,
        HHHHHHHHHHHHHHHHH,
        IIIIIIIIIIIIIIIII,
    > + Eq
    + PartialEq
{
}
```

whereas if it's a type definition
```rust
type AAAAAAAAAAAAA: BBBBBBBBBBBBBBB<
        CCCCCCCCCCCCCCCCC,
        DDDDDDDDDDDDDDDDD,
        EEEEEEEEEEEEEEEEE,
        FFFFFFFFFFFFFFFFF,
        GGGGGGGGGGGGGGGGG,
        HHHHHHHHHHHHHHHHH,
        IIIIIIIIIIIIIIIII,
    >;
```

I think the solution is to make sure the first trait bound goes to the next line
```rust
type AAAAAAAAAAAAA: 
    BBBBBBBBBBBBBBB<
        CCCCCCCCCCCCCCCCC,
        DDDDDDDDDDDDDDDDD,
        EEEEEEEEEEEEEEEEE,
        FFFFFFFFFFFFFFFFF,
        GGGGGGGGGGGGGGGGG,
        HHHHHHHHHHHHHHHHH,
        IIIIIIIIIIIIIIIII,
    >;
```

**Comment by ytmimi** [comment]
@johnhuichen I don't think we need to move the bound definition to the next line. My guess is that this is a `Shape` related issue. `Shape` is the type we use within the codebase to determine indentation. Given that things format correct with `v1` and don't format correctly with `v2` I suspect we must be using the wrong shape for `v2`.

It would be great if you could figure out what the difference is between the `v1` and `v2` codepaths. Let me know if you need help figuring out where in the codebase to get started.

**Comment by johnhuichen** [comment]
Actually I can probably do something about `retry_with_force_newline` there...

**Comment by johnhuichen** [comment]
@ytmimi after some digging through, I still think the problem is with inconsistent formatting. Since bounds is used by both trait and type, you can't without difficulty format traits as
```rust
trait SomeTrait:
    BBBBBBBBBBBBBBB<
        CCCCCCCCCCCCCCCCC,
        DDDDDDDDDDDDDDDDD,
        EEEEEEEEEEEEEEEEE,
        FFFFFFFFFFFFFFFFF,
        GGGGGGGGGGGGGGGGG,
        HHHHHHHHHHHHHHHHH,
        IIIIIIIIIIIIIIIII,
    >
{
}
```

while formatting type as
```rust
type SomeType: BBBBBBBBBBBBBBB<
    CCCCCCCCCCCCCCCCC,
    DDDDDDDDDDDDDDDDD,
    EEEEEEEEEEEEEEEEE,
    FFFFFFFFFFFFFFFFF,
>;
```

one of them needs to conform to the other.

**Comment by ytmimi** [comment]
@johnhuichen Thanks for continuing to look into this. To figure out how to proceed I've been looking throught the [Rust Style Guide](https://doc.rust-lang.org/nightly/style-guide/index.html).

The trait formatting looks right to me, and [the rules for formatting traits are described here](https://doc.rust-lang.org/nightly/style-guide/items.html#traits). 

```rust
trait SomeTrait:
    BBBBBBBBBBBBBBB<
        CCCCCCCCCCCCCCCCC,
        DDDDDDDDDDDDDDDDD,
        EEEEEEEEEEEEEEEEE,
        FFFFFFFFFFFFFFFFF,
        GGGGGGGGGGGGGGGGG,
        HHHHHHHHHHHHHHHHH,
        IIIIIIIIIIIIIIIII,
    >
{
}
```

[The associated type formatting rules are mentioned here](https://doc.rust-lang.org/nightly/style-guide/items.html#associated-types), but they seem a little under specified in this case and don't really tell us how to break the bound. The associated type docs mention to "Format associated types like type aliases", [though I don't think the type alias docs help us out much either](https://doc.rust-lang.org/nightly/style-guide/items.html#type-aliases).

[The Types and Bounds chapter of the guide](https://doc.rust-lang.org/nightly/style-guide/types.html#types-and-bounds) describes line breaking, and tells us to [break generic types following the rules for generics](https://doc.rust-lang.org/nightly/style-guide/types.html#:~:text=Break%20generic%20types%20following%20the%20rules%20for%20generics.).

[The rules for generics are here](https://doc.rust-lang.org/nightly/style-guide/items.html#generics), and based on my reading I'd expect the following to also be the correct formatting.

```rust
type SomeType: BBBBBBBBBBBBBBB<
    CCCCCCCCCCCCCCCCC,
    DDDDDDDDDDDDDDDDD,
    EEEEEEEEEEEEEEEEE,
    FFFFFFFFFFFFFFFFF,
>;
```

 That said, @rust-lang/style, I'd be curious to know if any of you have thoughts on this issue.

**Comment by johnhuichen** [comment]
@calebcartwright bounds implementation was  changed to fix
```rust
pub trait PrettyPrinter<'tcx>:
    Printer<
    'tcx,
    Error = fmt::Error,
    Path = Self,
    Region = Self,
    Type = Self,
    DynExistential = Self,
    Const = Self,
>
{
    //
}
```

But the fix broke formatting for type alias. I didn't find a straight forward way to fix type alias without breaking trait formatting. I can give it another try this weekend, but hopefully you see my dilemma
